tests/local.rs

```diff
@@ -0,0 +1,264 @@
+            port = prospective_port;
+            break
+        }
+    }
+
+    std::thread::spawn(move || {
+
+        let mut rt = tokio::runtime::Builder::new().enable_all().basic_scheduler().build().unwrap();
+        let server = start_server(port);
+        rt.block_on(server);
+    });
+    std::thread::sleep(std::time::Duration::from_secs(1));
+    port
+}
+
+
+
+/// Starts the fileserver
+async fn start_server(port: u16) {
+    let localhost = [0, 0, 0, 0];
+    let addr = (localhost, port);
+
+    const ASSETS_DIR: &str = "tests/test_html";
+    let assets_dir: PathBuf = PathBuf::from(ASSETS_DIR);
+    let routes = fileserver(assets_dir);
+    warp::serve(routes).run(addr).await
+}
+
+
+/// Serves files under this directory.
+fn fileserver(assets_dir: PathBuf) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
+    warp::get()
+        .and(warp::fs::dir(assets_dir))
+        .and(warp::path::end())
+}
+
+fn sample_page_url(port: u16) -> String {
+    format!("http://localhost:{}/sample_page.html", port)
+}
+
+
+async fn goto(mut c: Client, port: u16) -> Result<(), error::CmdError> {
+    let url = sample_page_url(port);
+    c.goto(&url).await?;
+    let current_url = c.current_url().await?;
+    assert_eq!(url.as_str(), current_url.as_str());
+    c.close().await
+}
+
+async fn new_window(mut c: Client, port: u16) -> Result<(), error::CmdError> {
+    let url = sample_page_url(port);
+    c.goto(&url).await?;
+    c.new_window(false).await?;
+    let handles = c.get_window_handles().await?;
+    assert_eq!(handles.len(), 2);
+    c.close().await
+}
+
+async fn new_window_switch(mut c: Client, _port: u16) -> Result<(), error::CmdError> {
+    let handle_1 = c.get_window_handle().await?;
+    c.new_window(false).await?;
+    let handle_2 = c.get_window_handle().await?;
+    assert_eq!(handle_1, handle_2, "After creating a new window, the session should not have switched to it");
+
+    let all_handles = c.get_window_handles().await?;
+    let new_window_handle = all_handles
+        .into_iter()
+        .find(|handle| handle != &handle_1)
+        .expect("Should find a differing handle");
+
+    c.switch_to_window(new_window_handle).await?;
+
+    let handle_3 = c.get_window_handle().await?;
+    assert_ne!(handle_3, handle_2, "After switching to a new window, the handle should differ now.");
+
+    c.close().await
+}
+
+
+async fn new_tab(mut c: Client, port: u16) -> Result<(), error::CmdError> {
+    let url = sample_page_url(port);
+    c.goto(&url).await?;
+    c.new_window(true).await?;
+    let handles = c.get_window_handles().await?;
+    assert_eq!(handles.len(), 2);
+    c.close().await
+}
+
+async fn close_window(mut c: Client, port: u16) -> Result<(), error::CmdError> {
+    let url = sample_page_url(port);
+    c.goto(&url).await?;
+    c.new_window(true).await?;
+    let handles = c.get_window_handles().await?;
+    assert_eq!(handles.len(), 2);
+    c.close_window().await?;
```

Comment:
It'd be good to have an assert after this that asserts what the _current_ window now is.

Reply:
Good catch here. After closing a window, it doesn't switch to a new window automatically. Calling `window()` again after closing the window will cause the webdriver to return a 404, because the window it currently is "switched to" no longer exists.

Do you think the `window()` command should return a `<Result<Option<Window>, _>` or just throw an error like it currently does?

Reply:
Scratch that - I've opted to keep the standard `Result<Window, _>` and made a new error variant to handle `NoSuchWindow` errors.

Reply:
I'm not sure what should happen here actually. In some sense I'm inclined to have `close_window` automatically switch to a different window if one exists, but that might be confusing too. `NoSuchWindow` seems pretty reasonable.

Reply:
There's decent documentation for that behavior if lib users are inclined to read it.
Its a niche set of commands that aren't used frequently, given how I'm the first to need them. I think slightly idiosyncratic behavior is acceptable here, given that the alternative would be for the close method to also make commands to get other windows and switch to one of them.